pkg/vault/contracts/ProtocolFeePercentagesProvider.sol

```diff
@@ -0,0 +1,149 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+pragma solidity ^0.8.24;
+
+import { SafeCast } from "@openzeppelin/contracts/utils/math/SafeCast.sol";
+
+import { IProtocolFeeController } from "@balancer-labs/v3-interfaces/contracts/vault/IProtocolFeeController.sol";
+import { IBasePoolFactory } from "@balancer-labs/v3-interfaces/contracts/vault/IBasePoolFactory.sol";
+import {
+    IProtocolFeePercentagesProvider
+} from "@balancer-labs/v3-interfaces/contracts/vault/IProtocolFeePercentagesProvider.sol";
+import {
+    IBalancerContractRegistry,
+    ContractType
+} from "@balancer-labs/v3-interfaces/contracts/standalone-utils/IBalancerContractRegistry.sol";
+import { IVault } from "@balancer-labs/v3-interfaces/contracts/vault/IVault.sol";
+
+import { SingletonAuthentication } from "./SingletonAuthentication.sol";
+
+contract ProtocolFeePercentagesProvider is IProtocolFeePercentagesProvider, SingletonAuthentication {
+    using SafeCast for uint256;
+
+    /**
+     * @dev Data structure to store default protocol fees by factory. Fee percentages are 18-decimal floating point
+     * numbers, so we know they fit in 64 bits, allowing the fees to be stored in a single slot.
+     *
+     * @param protocolSwapFee The protocol swap fee
+     * @param protocolYieldFee The protocol yield fee
+     * @param areFactoryFeesSet Flag indicating fees have been set (allows zero values)
+     */
+    struct FactoryProtocolFees {
+        uint64 protocolSwapFeePercentage;
+        uint64 protocolYieldFeePercentage;
+        bool areFactoryFeesSet;
+    }
+
+    IBalancerContractRegistry private immutable _trustedContractRegistry;
+    IProtocolFeeController private immutable _protocolFeeController;
+
+    uint256 private immutable _maxProtocolSwapFeePercentage;
+    uint256 private immutable _maxProtocolYieldFeePercentage;
+
+    // Factory address => FactoryProtocolFees
+    mapping(IBasePoolFactory => FactoryProtocolFees) private _factoryDefaultFeePercentages;
+
+    constructor(
+        IVault vault,
+        IProtocolFeeController protocolFeeController,
+        IBalancerContractRegistry trustedContractRegistry
+    ) SingletonAuthentication(vault) {
+        _protocolFeeController = protocolFeeController;
+        _trustedContractRegistry = trustedContractRegistry;
+
+        if (protocolFeeController.vault() != vault) {
+            revert WrongProtocolFeeControllerDeployment();
+        }
+
+        // These values are constant in the `ProtocolFeeController`.
+        (_maxProtocolSwapFeePercentage, _maxProtocolYieldFeePercentage) = protocolFeeController
+            .getMaximumProtocolFeePercentages();
```

Comment:
Unfortunately I don't think we'll migrate this once more in the short term. I'd just receive the constants and verify that they are correct in the fork test.

Reply:
They're public, so I can still get them; just not with this function. I think to avoid confusion I will not make any changes to the controller in this PR. (We've already changed it since V2, but not in this way.) Just what happens when PRs languish too long.